scala 3 "foreach: x =>" code not accepted

## Configuration ##
Here's my `.scalafmt.conf`:
```
version = "3.7.12"
runner.dialect = scala3
```

## Command-line parameters ##
scalafmt FmtTrouble.scala

## Steps

Given code like this (in ```FmtTrouble.scala```) :
```scala
@main def bug(): Unit =
  List(1,2,3) foreach: elem =>
    println(elem)

```

## Problem

Scalafmt outputs the following error:
```
$ scalafmt FmtTrouble.scala
FmtTrouble.scala:2: error: [dialect scala3] ; expected but => found
  List(1,2,3) foreach: elem =>
                            ^
```

## Expectation
As this code compile, would expect salafmt accept it :)
```
scala-cli FmtTrouble.scala
1
2
3
```
## Program versions
```
$ scala-cli --version
Scala CLI version: 1.0.4
Scala version (default): 3.3.0
$ scalafmt --version
scalafmt 3.7.12
```

--- a/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala
+++ b/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala
@@ -2286,46 +2286,9 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
         simpleExprRest(arguments, canApply = true, startPos = startPos)
       case _: Colon if canApply && dialect.allowFewerBraces =>
         val colonPos = tokenPos
-        // map:
-        val argsOpt = if (isEolAfterColonFewerBracesBody()) Some {
-          next()
-          blockExprOnIndent()
-        }
-        else
-          tryAhead(Try {
-            val paramPos = tokenPos
-
-            /**
-             * We need to handle param and then open indented region, otherwise only the block will
-             * be handles and any `.` will be accepted into the block:
-             * ```
-             * .map: a =>
-             * a+1
-             * .filter: x =>
-             * x > 2
-             * ```
-             * Without manual handling here, filter would be included for `(a+1).filter`
-             */
-            val param = simpleExpr(allowRepeated = false)
-            val contextFunction = token.is[ContextArrow]
-            if ((contextFunction || token.is[RightArrow]) && isIndentAfter())
-              convertToParamClause(param)(true, true).map { pc =>
-                val params = autoEndPos(paramPos)(pc)
-                next()
-                val trm = blockExprOnIndent()
-                autoEndPos(paramPos) {
-                  if (contextFunction)
-                    Term.ContextFunction(params, trm)
-                  else
-                    Term.Function(params, trm)
-                }
-              }
-            else None
-          }.getOrElse(None))
-
-        argsOpt match {
-          case Some(args) =>
-            val argClause = autoEndPos(colonPos)(Term.ArgClause(args :: Nil))
+        getFewerBracesArgOnColon() match {
+          case Some(arg) =>
+            val argClause = autoEndPos(colonPos)(Term.ArgClause(arg :: Nil))
             val arguments = addPos(Term.Apply(t, argClause))
             simpleExprRest(arguments, canApply = true, startPos = startPos)
           case _ => t
@@ -2336,6 +2299,44 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
       case _ =>
         t
     }
+  }
+
+  private def getFewerBracesArgOnColon(): Option[Term] = {
+    if (isEolAfterColonFewerBracesBody()) Some {
+      next()
+      blockExprOnIndent()
+    }
+    else
+      tryAhead(Try {
+        val paramPos = tokenPos
+
+        /**
+         * We need to handle param and then open indented region, otherwise only the block will be
+         * handles and any `.` will be accepted into the block:
+         * ```
+         * .map: a =>
+         * a+1
+         * .filter: x =>
+         * x > 2
+         * ```
+         * Without manual handling here, filter would be included for `(a+1).filter`
+         */
+        val param = simpleExpr(allowRepeated = false)
+        val contextFunction = token.is[ContextArrow]
+        if ((contextFunction || token.is[RightArrow]) && isIndentAfter())
+          convertToParamClause(param)(true, true).map { pc =>
+            val params = autoEndPos(paramPos)(pc)
+            next()
+            val trm = blockExprOnIndent()
+            autoEndPos(paramPos) {
+              if (contextFunction)
+                Term.ContextFunction(params, trm)
+              else
+                Term.Function(params, trm)
+            }
+          }
+        else None
+      }.getOrElse(None))
   }
 
   private def argumentExprsOrPrefixExpr(location: Location): Term = {

--- a/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala
+++ b/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala
@@ -2057,17 +2057,25 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
         ctx.reduceStack(base, rhsK, rhsEndK, Some(op))
       }
 
-      def getNextRhs(op: Term.Name, targs: Type.ArgClause): Term = {
+      def getNextRhs(op: Term.Name, targs: Type.ArgClause) =
+        getNextRhsWith(op, targs, argumentExprsOrPrefixExpr(PostfixStat))
+
+      def getNextRhsWith(op: Term.Name, targs: Type.ArgClause, rhs: Term) = {
         val lhs = getPrevLhs(op)
         val wrap = (lhs eq rhs0) && lhs.startTokenPos != startPos
         val lhsExt = if (wrap) atPosWithBody(startPos, Term.Tuple(lhs :: Nil), rhsEndK) else lhs
         ctx.push(ctx.UnfinishedInfix(lhsExt, op, targs))
-        argumentExprsOrPrefixExpr(PostfixStat)
+        Right(rhs)
       }
 
-      def getPostfix(op: Term.Name): Term = {
+      def getPostfix(op: Term.Name, targs: Type.ArgClause) = {
+        // Infix chain has ended with a postfix expression.
+        // This never happens in the running example.
+        if (targs.nonEmpty)
+          syntaxError("type application is not allowed for postfix operators", at = token)
         val finQual = getPrevLhs(op)
-        atPos(getLhsStartPos(finQual), op)(Term.Select(finQual, op))
+        val term: Term = atPos(getLhsStartPos(finQual), op)(Term.Select(finQual, op))
+        Left(term)
       }
 
       def getPostfixOrNextRhs(op: Term.Name): Either[Term, Term] = {
@@ -2082,14 +2090,10 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
           else isExprIntro(token, tokenPos)) {
           // Infix chain continues, so we need to reduce the stack.
           // In the running example, base = List(), rhsK = [a].
-          Right(getNextRhs(op, targs)) // [a]
+          getNextRhs(op, targs) // [a]
           // afterwards, ctx.stack = List([a +])
         } else {
-          // Infix chain has ended with a postfix expression.
-          // This never happens in the running example.
-          if (targs.nonEmpty)
-            syntaxError("type application is not allowed for postfix operators", at = token)
-          Left(getPostfix(op))
+          getPostfix(op, targs)
         }
       }
 
@@ -2106,7 +2110,7 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
           Some(Right(matchClause(lhs, getLhsStartPos(lhs))))
         case _: LF if dialect.allowInfixOperatorAfterNL =>
           tryGetNextInfixOpIfLeading(startPos)(Term.Name.apply).map { op =>
-            Right(getNextRhs(op, autoPos(Type.ArgClause(Nil))))
+            getNextRhs(op, autoPos(Type.ArgClause(Nil)))
           }
         case _ => None
       }

--- a/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala
+++ b/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala
@@ -2093,7 +2093,9 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
           getNextRhs(op, targs) // [a]
           // afterwards, ctx.stack = List([a +])
         } else {
-          getPostfix(op, targs)
+          (if (token.is[Colon] && dialect.allowFewerBraces) getFewerBracesArgOnColon() else None)
+            .map(arg => getNextRhsWith(op, targs, arg))
+            .getOrElse(getPostfix(op, targs))
         }
       }
 

--- a/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/FewerBracesSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/FewerBracesSuite.scala
@@ -268,7 +268,7 @@ class FewerBracesSuite extends BaseDottySuite {
          |    file
          |""".stripMargin,
       assertLayout = Some(
-        """|def O = credentials.++ {
+        """|def O = credentials ++ {
            |  val file = Path.userHome / ".credentials"
            |  file
            |}
@@ -280,8 +280,10 @@ class FewerBracesSuite extends BaseDottySuite {
         tname("O"),
         Nil,
         None,
-        Term.Apply(
-          Term.Select(tname("credentials"), tname("++")),
+        Term.ApplyInfix(
+          tname("credentials"),
+          tname("++"),
+          Nil,
           Term.Block(
             List(
               Defn.Val(
@@ -712,11 +714,23 @@ class FewerBracesSuite extends BaseDottySuite {
         |  List(1,2,3) foreach: x =>
         |    println(x)
         |""".stripMargin
-    val error =
-      """|<input>:2: error: ; expected but => found
-         |  List(1,2,3) foreach: x =>
-         |                         ^""".stripMargin
-    runTestError[Stat](code, error)
+    val layout = "object MyApp { List(1, 2, 3) foreach (x => println(x)) }"
+    val tree = Defn.Object(
+      Nil,
+      tname("MyApp"),
+      tpl(
+        Term.ApplyInfix(
+          Term.Apply(tname("List"), List(int(1), int(2), int(3))),
+          tname("foreach"),
+          Nil,
+          Term.Function(
+            List(tparam("x", None)),
+            Term.Apply(tname("println"), List(tname("x")))
+          ) :: Nil
+        ) :: Nil
+      )
+    )
+    runTestAssert[Stat](code, Some(layout))(tree)
   }
 
   test("#3319 lambda simple lhs") {
@@ -725,11 +739,23 @@ class FewerBracesSuite extends BaseDottySuite {
         |  ids foreach: x =>
         |    println(x)
         |""".stripMargin
-    val error =
-      """|<input>:2: error: ; expected but => found
-         |  ids foreach: x =>
-         |                 ^""".stripMargin
-    runTestError[Stat](code, error)
+    val layout = "object MyApp { ids foreach (x => println(x)) }"
+    val tree = Defn.Object(
+      Nil,
+      tname("MyApp"),
+      tpl(
+        Term.ApplyInfix(
+          tname("ids"),
+          tname("foreach"),
+          Nil,
+          Term.Function(
+            List(tparam("x", None)),
+            Term.Apply(tname("println"), List(tname("x")))
+          ) :: Nil
+        ) :: Nil
+      )
+    )
+    runTestAssert[Stat](code, Some(layout))(tree)
   }
 
   test("#3319 lambda 'chained'") {
@@ -740,11 +766,36 @@ class FewerBracesSuite extends BaseDottySuite {
         |    map: x =>
         |      bar(x)
         |""".stripMargin
-    val error =
-      """|<input>:2: error: ; expected but => found
-         |  ids map: x =>
-         |             ^""".stripMargin
-    runTestError[Stat](code, error)
+    val layout =
+      """|object MyApp {
+         |  ids map (x => foo(x))
+         |  map(x => bar(x))
+         |}""".stripMargin
+    val tree = Defn.Object(
+      Nil,
+      tname("MyApp"),
+      tpl(
+        List(
+          Term.ApplyInfix(
+            tname("ids"),
+            tname("map"),
+            Nil,
+            Term.Function(
+              List(tparam("x", None)),
+              Term.Apply(tname("foo"), List(tname("x")))
+            ) :: Nil
+          ),
+          Term.Apply(
+            tname("map"),
+            Term.Function(
+              List(tparam("x", None)),
+              Term.Apply(tname("bar"), List(tname("x")))
+            ) :: Nil
+          )
+        )
+      )
+    )
+    runTestAssert[Stat](code, Some(layout))(tree)
   }
 
   test("#3319 lambda block") {
@@ -754,11 +805,34 @@ class FewerBracesSuite extends BaseDottySuite {
         |      foo(x)
         |      bar(x)
         |""".stripMargin
-    val error =
-      """|<input>:2: error: ; expected but => found
-         |  ids map: x =>
-         |             ^""".stripMargin
-    runTestError[Stat](code, error)
+    val layout =
+      """|object MyApp {
+         |  ids map (x => {
+         |    foo(x)
+         |    bar(x)
+         |  })
+         |}""".stripMargin
+    val tree = Defn.Object(
+      Nil,
+      tname("MyApp"),
+      tpl(
+        Term.ApplyInfix(
+          tname("ids"),
+          tname("map"),
+          Nil,
+          Term.Function(
+            List(tparam("x", None)),
+            Term.Block(
+              List(
+                Term.Apply(tname("foo"), List(tname("x"))),
+                Term.Apply(tname("bar"), List(tname("x")))
+              )
+            )
+          ) :: Nil
+        ) :: Nil
+      )
+    )
+    runTestAssert[Stat](code, Some(layout))(tree)
   }
 
   test("#3319 function complex lhs") {
@@ -767,13 +841,15 @@ class FewerBracesSuite extends BaseDottySuite {
         |  List(1,2,3) foreach:
         |    println
         |""".stripMargin
-    val layout = "object MyApp { List(1, 2, 3).foreach(println) }"
+    val layout = "object MyApp { List(1, 2, 3) foreach println }"
     val tree = Defn.Object(
       Nil,
       tname("MyApp"),
       tpl(
-        Term.Apply(
-          Term.Select(Term.Apply(tname("List"), List(int(1), int(2), int(3))), tname("foreach")),
+        Term.ApplyInfix(
+          Term.Apply(tname("List"), List(int(1), int(2), int(3))),
+          tname("foreach"),
+          Nil,
           List(tname("println"))
         ) :: Nil
       )
@@ -787,13 +863,15 @@ class FewerBracesSuite extends BaseDottySuite {
         |  ids foreach:
         |    println
         |""".stripMargin
-    val layout = "object MyApp { ids.foreach(println) }"
+    val layout = "object MyApp { ids foreach println }"
     val tree = Defn.Object(
       Nil,
       tname("MyApp"),
       tpl(
-        Term.Apply(
-          Term.Select(tname("ids"), tname("foreach")),
+        Term.ApplyInfix(
+          tname("ids"),
+          tname("foreach"),
+          Nil,
           List(tname("println"))
         ) :: Nil
       )
@@ -811,7 +889,7 @@ class FewerBracesSuite extends BaseDottySuite {
         |""".stripMargin
     val layout =
       """|object MyApp {
-         |  ids.map(foo)
+         |  ids map foo
          |  map(bar)
          |}""".stripMargin
     val tree = Defn.Object(
@@ -819,10 +897,7 @@ class FewerBracesSuite extends BaseDottySuite {
       tname("MyApp"),
       tpl(
         List(
-          Term.Apply(
-            Term.Select(tname("ids"), tname("map")),
-            List(tname("foo"))
-          ),
+          Term.ApplyInfix(tname("ids"), tname("map"), Nil, List(tname("foo"))),
           Term.Apply(tname("map"), List(tname("bar")))
         )
       )

--- a/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/FewerBracesSuite.scala
+++ b/tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/FewerBracesSuite.scala
@@ -277,30 +277,27 @@ class FewerBracesSuite extends BaseDottySuite {
     )(
       Defn.Def(
         Nil,
-        Term.Name("O"),
-        Nil,
+        tname("O"),
         Nil,
         None,
         Term.Apply(
-          Term.Select(Term.Name("credentials"), Term.Name("++")),
-          List(
-            Term.Block(
-              List(
-                Defn.Val(
+          Term.Select(tname("credentials"), tname("++")),
+          Term.Block(
+            List(
+              Defn.Val(
+                Nil,
+                List(Pat.Var(tname("file"))),
+                None,
+                Term.ApplyInfix(
+                  Term.Select(tname("Path"), tname("userHome")),
+                  tname("/"),
                   Nil,
-                  List(Pat.Var(Term.Name("file"))),
-                  None,
-                  Term.ApplyInfix(
-                    Term.Select(Term.Name("Path"), Term.Name("userHome")),
-                    Term.Name("/"),
-                    Nil,
-                    List(Lit.String(".credentials"))
-                  )
-                ),
-                Term.Name("file")
-              )
+                  List(str(".credentials"))
+                )
+              ),
+              tname("file")
             )
-          )
+          ) :: Nil
         )
       )
     )
@@ -707,6 +704,130 @@ class FewerBracesSuite extends BaseDottySuite {
          |}
          |""".stripMargin
     assertNoDiff(parseStat(code, dialect).reprint, layout)
+  }
+
+  test("#3319 lambda complex lhs") {
+    val code =
+      """object MyApp:
+        |  List(1,2,3) foreach: x =>
+        |    println(x)
+        |""".stripMargin
+    val error =
+      """|<input>:2: error: ; expected but => found
+         |  List(1,2,3) foreach: x =>
+         |                         ^""".stripMargin
+    runTestError[Stat](code, error)
+  }
+
+  test("#3319 lambda simple lhs") {
+    val code =
+      """object MyApp:
+        |  ids foreach: x =>
+        |    println(x)
+        |""".stripMargin
+    val error =
+      """|<input>:2: error: ; expected but => found
+         |  ids foreach: x =>
+         |                 ^""".stripMargin
+    runTestError[Stat](code, error)
+  }
+
+  test("#3319 lambda 'chained'") {
+    val code =
+      """object MyApp:
+        |  ids map: x =>
+        |      foo(x)
+        |    map: x =>
+        |      bar(x)
+        |""".stripMargin
+    val error =
+      """|<input>:2: error: ; expected but => found
+         |  ids map: x =>
+         |             ^""".stripMargin
+    runTestError[Stat](code, error)
+  }
+
+  test("#3319 lambda block") {
+    val code =
+      """object MyApp:
+        |  ids map: x =>
+        |      foo(x)
+        |      bar(x)
+        |""".stripMargin
+    val error =
+      """|<input>:2: error: ; expected but => found
+         |  ids map: x =>
+         |             ^""".stripMargin
+    runTestError[Stat](code, error)
+  }
+
+  test("#3319 function complex lhs") {
+    val code =
+      """object MyApp:
+        |  List(1,2,3) foreach:
+        |    println
+        |""".stripMargin
+    val layout = "object MyApp { List(1, 2, 3).foreach(println) }"
+    val tree = Defn.Object(
+      Nil,
+      tname("MyApp"),
+      tpl(
+        Term.Apply(
+          Term.Select(Term.Apply(tname("List"), List(int(1), int(2), int(3))), tname("foreach")),
+          List(tname("println"))
+        ) :: Nil
+      )
+    )
+    runTestAssert[Stat](code, Some(layout))(tree)
+  }
+
+  test("#3319 function simple lhs") {
+    val code =
+      """object MyApp:
+        |  ids foreach:
+        |    println
+        |""".stripMargin
+    val layout = "object MyApp { ids.foreach(println) }"
+    val tree = Defn.Object(
+      Nil,
+      tname("MyApp"),
+      tpl(
+        Term.Apply(
+          Term.Select(tname("ids"), tname("foreach")),
+          List(tname("println"))
+        ) :: Nil
+      )
+    )
+    runTestAssert[Stat](code, Some(layout))(tree)
+  }
+
+  test("#3319 function 'chained'") {
+    val code =
+      """object MyApp:
+        |  ids map:
+        |      foo
+        |    map:
+        |      bar
+        |""".stripMargin
+    val layout =
+      """|object MyApp {
+         |  ids.map(foo)
+         |  map(bar)
+         |}""".stripMargin
+    val tree = Defn.Object(
+      Nil,
+      tname("MyApp"),
+      tpl(
+        List(
+          Term.Apply(
+            Term.Select(tname("ids"), tname("map")),
+            List(tname("foo"))
+          ),
+          Term.Apply(tname("map"), List(tname("bar")))
+        )
+      )
+    )
+    runTestAssert[Stat](code, Some(layout))(tree)
   }
 
 }

